source/adaptive_load/metrics_plugin_impl.cc

```diff
@@ -0,0 +1,221 @@
+ * @param result_name The name of the desired Result, e.g. "global".
+ *
+ * @return StatusOr<Result> The Result with the specified name if found, otherwise an error Status.
+ */
+absl::StatusOr<nighthawk::client::Result>
+GetResult(const nighthawk::client::Output& nighthawk_output, absl::string_view result_name) {
+  for (const nighthawk::client::Result& result : nighthawk_output.results()) {
+    if (result.name() == result_name) {
+      return result;
+    }
+  }
+  return absl::InternalError(
+      absl::StrCat("Result '", result_name, "' not found in Nighthawk output."));
+}
+
+/**
+ * Returns the value of the counter with the given name within a Result proto.
+ *
+ * @param result A Result proto taken from a Nighthawk Output proto.
+ * @param counter_name The name of the counter, e.g. "benchmark.http_2xx".
+ *
+ * @return StatusOr<uint32_t> The counter value if found, otherwise an error Status.
+ */
+absl::StatusOr<uint32_t> GetCounter(const nighthawk::client::Result& result,
+                                    absl::string_view counter_name) {
+  for (const nighthawk::client::Counter& counter : result.counters()) {
+    if (counter.name() == counter_name) {
+      return counter.value();
+    }
+  }
+  return absl::InternalError(
+      absl::StrCat("Counter '", counter_name, "' not found in Result proto."));
+}
+
+/**
+ * Finds a Statistic proto with the given id within a Result proto.
+ *
+ * @param result A Result proto taken from a Nighthawk Output proto.
+ * @param statistic_id The name of the desired Statistic, e.g.
+ * "benchmark_http_client.request_to_response".
+ *
+ * @return StatusOr<Statistic> The Statistic with the specified id if found, otherwise an error
+ * Status.
+ */
+absl::StatusOr<nighthawk::client::Statistic> GetStatistic(const nighthawk::client::Result& result,
+                                                          absl::string_view statistic_id) {
+  for (const nighthawk::client::Statistic& statistic : result.statistics()) {
+    if (statistic.id() == statistic_id) {
+      return statistic;
+    }
+  }
+  return absl::InternalError(
+      absl::StrCat("Statistic '", statistic_id, "' not found in Result proto."));
+}
+
+/**
+ * Extracts a value from a StatusOr into a variable if OK; otherwise appends the error message to a
+ * vector. static_casts the value to fit the variable. Ignores Status code values. For use in a
+ * constructor that stores a vector of errors during initialization.
+ *
+ * @param status_or The StatusOr to extract. The value must be static_castable to the type of
+ * |value|.
+ * @param value The variable where the value should be extracted. Only written if the status is OK.
+ * @param errors A place to accumulate error messages. If the status is not OK, its message is
+ * appended here.
+ *
+ * @return bool True if the status was OK.
+ */
+template <typename T, typename U>
+bool ExtractValueOrRecordError(absl::StatusOr<T> status_or, U& value,
```

Comment:
I understand the motivation behind this template (removal of code duplication), however we should also consider the complexity it adds for the reader, it is both a redirection and a meta code just to process errors.

While the code without this template would contain a few more lines on each error check, what we would get in return is lower barrier of adoption, especially for potential contributors who aren't very versed with c++. Let's try to keep the code as simple as possible even if it means having a few repeated lines.

Reply:
Done